### PR TITLE
[DR-2107] Remove google billing account id and project from Azure resources

### PR DESCRIPTION
**Pull request by samanehsan**
Changes:
- Make the `billingAccountId` optional in the create profile API request (but still check that it is specified for gcp billing profiles)
- Make the `billingAccountId` field nullable in the database
- Add a step ([VerifyBillingAccountAccessStep](https://github.com/DataBiosphere/jade-data-repo/pull/1097/files?authenticity_token=0ZOOnGD%2BVL8D3HUpjqOVYX8aKEAyscYgi%2FrvLCHoS8XktTzpoNlJhhE6oVhcXcZc%2BkVwYviSN0tXB5SemTg6dw%3D%3D&file-filters%5B%5D=.java&file-filters%5B%5D=.xml#diff-b927632e2fe372069c1d8b3a6e267d669ee0b7d78c1aa3ccd95978f3066408e1)) to check that the data repo has access to the gcp billing profile (was previously checked as part of AuthorizeBillingProfileUseStep [here](https://github.com/DataBiosphere/jade-data-repo/pull/1097/files?authenticity_token=0ZOOnGD%2BVL8D3HUpjqOVYX8aKEAyscYgi%2FrvLCHoS8XktTzpoNlJhhE6oVhcXcZc%2BkVwYviSN0tXB5SemTg6dw%3D%3D&file-filters%5B%5D=.java&file-filters%5B%5D=.xml#diff-1f89b17e409cc8e853e0548d3e53ccb9b1c706d6dcc7f761ab6b43bf55236b30))
- Add a step ([VerifyDeployedApplicationAccessStep](https://github.com/DataBiosphere/jade-data-repo/pull/1097/files?authenticity_token=0ZOOnGD%2BVL8D3HUpjqOVYX8aKEAyscYgi%2FrvLCHoS8XktTzpoNlJhhE6oVhcXcZc%2BkVwYviSN0tXB5SemTg6dw%3D%3D&file-filters%5B%5D=.java&file-filters%5B%5D=.xml#diff-e5aa9ef15a75822b3c7387838514eba8ae9532bf2138c9821fe1140e6f3aca79)) to verify access to the Azure deployed application
- Do not set up a google project for Azure resources
- Make the project_resource_id nullable for datasets and snapshots in the database
- Update retrieving the snapshotProject for DRS lookups to handle Azure snapshots no longer having a google data project

### CODE REVIEW

File: src/main/java/bio/terra/service/dataset/DatasetDao.java
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/DatasetDao.java:497]: I'm not too familiar with this code, but is there a reason to keep this field around once this change is complete? Are there existing azure datasets in TDR that we need to support for backward compatibility reasons? I would expect to see the field go away completely.

Comment (samanehsan) [reply src/main/java/bio/terra/service/dataset/DatasetDao.java:497]: I was thinking backwards-compatibility like you mentioned (but mainly for GCP datasets). I don't think we have any azure datasets that aren't test data.

Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/dataset/DatasetDao.java:497]: Oh, I see, I thought this part was azure only. That makes sense, it's still used for GCP datasets. Before everything used this field, now its use varies depending on the platform.

File: src/main/java/bio/terra/service/profile/ProfileService.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/profile/ProfileService.java]: Just want to confirm: we were doing this in a separate step and now this combined into one...is that correct?

File: src/main/resources/api/data-repository-openapi.yaml
Comment (nmalfroy) [inline src/main/resources/api/data-repository-openapi.yaml:2671]: Biller probably shouldn't be required anymore either since we only use it for Google? We may eventually want to add `cloudPlatform` but that's a breaking change so maybe we hold off on that

**Review by nmalfroy — APPROVED** [review]
Looks great!  Just a quick sanity check question on my end

File: src/main/java/bio/terra/service/profile/ProfileService.java
Comment (samanehsan) [reply src/main/java/bio/terra/service/profile/ProfileService.java]: We weren't doing this check for Azure, so I added it in here. Do you think it makes more sense to put those in separate steps?

File: src/main/resources/api/data-repository-openapi.yaml
Comment (samanehsan) [reply src/main/resources/api/data-repository-openapi.yaml:2671]: Good point!

Comment (samanehsan) [reply src/main/resources/api/data-repository-openapi.yaml:2671]: Oh I see the biller is being used for Azure here: https://github.com/DataBiosphere/jade-data-repo/blob/6b0e9faf17a7f9264944c235c0b5ddba14eb5417/src/main/java/bio/terra/service/filedata/flight/ingest/IngestBuildAndWriteScratchLoadFileAzureStep.java#L80

Could we replace this with something else?

File: src/main/resources/db/changesets/20211015_allownullbillingaccountandprojectresourceid.yaml
Comment (snf2ye) [inline src/main/resources/db/changesets/20211015_allownullbillingaccountandprojectresourceid.yaml:6]: Given that we're removing these constraints, do you think we sufficiently confirm that either a google project resource or an azure storage account is defined?

File: src/main/java/bio/terra/service/dataset/flight/ingest/DatasetIngestFlight.java
Comment (snf2ye) [inline src/main/java/bio/terra/service/dataset/flight/ingest/DatasetIngestFlight.java]: > // TODO: We can get rid of this once https://broadworkbench.atlassian.net/browse/DR-2107
      // is complete

I think we can remove this comment. Thanks to @tlangs for refreshing my memory. The original thought was that we were only adding the AuthorizeBillingProfileUseStep to accommodate that fact that Azure datasets still referenced a Google project. But, since this is now also checking for authorization for the storage account, I think it's okay as is.

File: src/main/java/bio/terra/service/dataset/flight/ingest/DatasetIngestFlight.java
Comment (snf2ye) [inline src/main/java/bio/terra/service/dataset/flight/ingest/DatasetIngestFlight.java]: We can also get rid of this comment.

File: src/main/java/bio/terra/service/snapshot/flight/create/SnapshotCreateFlight.java
Comment (snf2ye) [inline src/main/java/bio/terra/service/snapshot/flight/create/SnapshotCreateFlight.java]: Line 100-101, there is another DR-2107 comment that can be removed.

File: src/main/java/bio/terra/service/profile/ProfileService.java
Comment (snf2ye) [reply src/main/java/bio/terra/service/profile/ProfileService.java]: I think separate steps would follow the precedence set in many of our other Azure/GCP steps.

File: src/main/java/bio/terra/service/profile/ProfileRequestValidator.java
Comment (snf2ye) [inline src/main/java/bio/terra/service/profile/ProfileRequestValidator.java:35]: Should we instead check that either a billing profile or a storage account is specified? And I would think we still want to validate the billing account id if it is provided.

**Review by snf2ye — COMMENTED** [review]
Thanks for cleaning up a few of my lingering code comments 😅

File: src/main/java/bio/terra/service/profile/ProfileRequestValidator.java
Comment (samanehsan) [reply src/main/java/bio/terra/service/profile/ProfileRequestValidator.java:35]: I moved this validation to line 75, so we validate the billing account id for GCP billing profiles. I think the checks in `isValidCloudPlatform` are sufficient...what do you think?

Comment (snf2ye) [reply src/main/java/bio/terra/service/profile/ProfileRequestValidator.java:35]: Oh, I misread what this validation was originally doing. This all looks good to me!

**Comment by samanehsan** [comment]
The test failure is from the develop branch

**Review by snf2ye — APPROVED** [review]

File: src/main/java/bio/terra/service/dataset/flight/create/DatasetCreateFlight.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/flight/create/DatasetCreateFlight.java]: I think you only want to do this step when you create a billing profile. Same with the `VerifyBillingAccountAccessStep` step above. Those steps look like they ensure that the user is allowed to use those cloud resources.  dataset creation should only be checking link permissions to the TDR spend profile.

File: src/main/java/bio/terra/service/dataset/flight/ingest/DatasetIngestFlight.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/dataset/flight/ingest/DatasetIngestFlight.java]: Same...I don't think you want this step here

File: src/main/java/bio/terra/service/profile/flight/VerifyBillingAccountAccessStep.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/profile/flight/VerifyBillingAccountAccessStep.java:15]: These validation steps should probably have some javadoc (and explain that they are explicitly doing validation against the cloud, not Sam)

File: src/main/java/bio/terra/service/snapshot/SnapshotDao.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/snapshot/SnapshotDao.java]: one of the main reasons to roll this onto a select statement in the first place was to reduce the amount of calls DRS lookups need to make to the db.  You could actually leave this by just doing a left join to the snapshot project table
```suggestion
            "(SELECT jsonb_agg(ds)\n"
            + "  FROM (SELECT d.id, d.name, p.profile_id as \"profileId\", p.google_project_id as \"dataProject\"\n"
            + "        FROM snapshot_source ss JOIN dataset d ON  ss.dataset_id = d.id"
            + "           LEFT JOIN project_resource p ON d.project_resource_id = p.id"
            + "        WHERE snapshot.id = ss.snapshot_id) ds)"
            + "  AS dataset_sources, "
            + "FROM snapshot "
            + "LEFT JOIN project_resource ON snapshot.project_resource_id = project_resource.id "
            + "WHERE snapshot.id = :id";
```

File: src/main/java/bio/terra/service/snapshot/SnapshotDao.java
Comment (nmalfroy) [inline src/main/java/bio/terra/service/snapshot/SnapshotDao.java]: Again, I'd keep the existing structure to reduce the number of queries

**Review by nmalfroy — CHANGES_REQUESTED** [review]
Looking good!  A couple of of things that I think we need to make sure before we merge:
- We shouldn't be verifying access to billing/application in each flight.  That really should only be in profile creation
- We should keep the existing query structure for perf reasons (also, I think it'll simplify your dao changes)

File: src/main/java/bio/terra/service/dataset/flight/create/DatasetCreateFlight.java
Comment (samanehsan) [reply src/main/java/bio/terra/service/dataset/flight/create/DatasetCreateFlight.java]: Ok got it! Thanks for clarifying!

File: src/main/java/bio/terra/service/snapshot/SnapshotDao.java
Comment (samanehsan) [reply src/main/java/bio/terra/service/snapshot/SnapshotDao.java]: That makes sense! That's so much nicer!!!

Comment (samanehsan) [reply src/main/java/bio/terra/service/snapshot/SnapshotDao.java]: We still want to detect the cloud provider right?

**Comment by samanehsan** [comment]
@nmalfroy I made the changes you requested! Let me know what you think!

**Review by nmalfroy — APPROVED** [review]
Thanks for making the changes!

**Pull request merged**
